File: Internet-services-architecture-and-development/Esercitazione3/Soluzioni/rgrep_20231129/client-lpd-connreuse.c
#define _POSIX_C_SOURCE 200809L
#include <errno.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include "utils.h"


int main(int argc, char **argv)
{
        uint8_t buffer[2048], len[2];
        char filename[512], stringa[512];
        int filename_len, stringa_len;
        int sd, err, nread;
        struct addrinfo hints, *ptr, *res;

        if (argc != 3) {
                fprintf(stderr, "Sintassi: largest_expenses server porta\n");
                exit(EXIT_FAILURE);
        }

	/* Ignoro SIGPIPE */
	signal(SIGPIPE, SIG_IGN);

        memset(&hints, 0, sizeof(hints));
        hints.ai_family   = AF_UNSPEC;
        hints.ai_socktype = SOCK_STREAM;

        err = getaddrinfo(argv[1], argv[2], &hints, &res);
        if (err != 0) {
                fprintf(stderr, "Errore risoluzione nome: %s\n", gai_strerror(err));
                exit(EXIT_FAILURE);
        }

        /* connessione con fallback */
        for (ptr=res; ptr != NULL; ptr=ptr->ai_next) {
                sd = socket(ptr->ai_family, ptr->ai_socktype, ptr->ai_protocol);
                /* se la socket fallisce, passo all'indirizzo successivo */
                if (sd < 0)
                        continue;

                /* se la connect va a buon fine, esco dal ciclo */
                if (connect(sd, ptr->ai_addr, ptr->ai_addrlen) == 0)
                        break;

                /* altrimenti, chiudo la socket e passo all'indirizzo
                 * successivo */
                close(sd);
        }

        /* controllo che effettivamente il client sia connesso */
        if (ptr == NULL) {
                fprintf(stderr, "Errore di connessione!\n");
                exit(EXIT_FAILURE);
        }

        /* a questo punto, posso liberare la memoria allocata da getaddrinfo */
        freeaddrinfo(res);

        
        for (;;) {
		/* Leggo filename */
                printf("Digita il filename di interesse ('fine' per terminare):\n");
                if (scanf("%s", filename) != 1) {
                        perror("scanf");
                        exit(EXIT_FAILURE);
                }

                if (strcmp(filename, "fine") == 0) {
                        break;
                }

		/* Leggo stringa */
                printf("Digita la stringa di interesse:\n");
                if (scanf("%s", stringa) != 1) {
                        perror("scanf");
                        exit(EXIT_FAILURE);
                }

                /* Trasmetto filename */
                filename_len = strlen(filename);
                if (filename_len > UINT16_MAX) {
                        fprintf(stderr, "filename troppo grande (massimo %d byte)!\n", UINT16_MAX);
                        exit(EXIT_FAILURE);
                }

                /* Codifico lunghezza stringa1 come intero unsigned a 16 bit in formato
                 * big endian (AKA network byte order) */
                len[0] = (uint8_t)((filename_len & 0xFF00) >> 8);
                len[1] = (uint8_t)(filename_len & 0xFF);

                /* Invio lunghezza filename */
                if (write_all(sd, len, 2) < 0) {
                        perror("write");
                        exit(EXIT_FAILURE);
                }

                /* Invio filename */
                if (write_all(sd, filename, filename_len) < 0) {
                        perror("write");
                        exit(EXIT_FAILURE);
                }

                /* Trasmetto stringa */
                stringa_len = strlen(stringa);
                if (stringa_len > UINT16_MAX) {
                        fprintf(stderr, "stringa troppo grande (massimo %d byte)!\n", UINT16_MAX);
                        exit(EXIT_FAILURE);
                }

                /* Codifico lunghezza stringa1 come intero unsigned a 16 bit in formato
                 * big endian (AKA network byte order) */
                len[0] = (uint8_t)((stringa_len & 0xFF00) >> 8);
                len[1] = (uint8_t)(stringa_len & 0xFF);

                /* Invio lunghezza stringa */
                if (write_all(sd, len, 2) < 0) {
                        perror("write");
                        exit(EXIT_FAILURE);
                }

                /* Invio stringa */
                if (write_all(sd, stringa, stringa_len) < 0) {
                        perror("write");
                        exit(EXIT_FAILURE);
                }

                /* Leggo lunghezza risposta */
                if (read_all(sd, len, 2) < 0) {
                        perror("read");
                        exit(EXIT_FAILURE);
                }

                size_t risposta_len = ((size_t)len[0]) << 8 | (size_t)len[1];
                size_t to_read = risposta_len;

                /* Stampo contenuto risposta a video */
                while (to_read > 0) {
                        size_t bufsize = sizeof(buffer);
                        size_t sz = (to_read < bufsize) ? to_read : bufsize;

                        nread = read(sd, buffer, sz);
                        if (nread < 0) {
                                perror("read");
                                exit(EXIT_FAILURE);
                        }

                        if (write_all(1, buffer, nread) < 0) {
                                perror("write");
                                exit(EXIT_FAILURE);
                        }

                        to_read -= nread;
                }

                /* Stampo un \n prima di terminare */
                if (write(1, "\n", 1) < 0) {
                        perror("write");
                        exit(EXIT_FAILURE);
                }
        }

        /* Chiudo la socket */
        close(sd);

        return 0;
}

File: III - Architettura e sviluppo servizi internet/Esercizi/Esercitazione3/Soluzioni/rgrep_20231129/ClientLPDConnreuse.java
import java.io.*;
import java.net.*;


public class ClientLPDConnreuse {
    public static void main(String[] argv) {

        if (argv.length != 2) {
            System.err.println("Uso corretto: java ClientLPDConnreuse server porta\n");
            System.exit(1);
        }

        try{
            var fromUser = new BufferedReader(new InputStreamReader(System.in));

            var s = new Socket(argv[0], Integer.parseInt(argv[1]));
            var fromServer = s.getInputStream();
            var toServer  = s.getOutputStream();
            var len = new byte[2];

            do {
                System.out.println("Inserisci nomefile: ");
                var nomefile = fromUser.readLine();

                if (nomefile.equals("fine")) {
                    break;
                }

                System.out.println("Inserisci stringa: ");
                var stringa = fromUser.readLine();

                // effettuo encoding di nomefile in UTF-8 e calcolo la
                // dimensione in byte del buffer risultante
                var nomefile_utf8 = nomefile.getBytes("UTF-8");
                var nomefile_len = nomefile_utf8.length;

                // effettuo encoding della dimensione del buffer nomefile_utf8
                // in un unsigned a 16 bit con ordine big endian
                len[0] = (byte)((nomefile_len & 0xFF00) >> 8);
                len[1] = (byte)(nomefile_len & 0xFF);

                // trasmetto nomefile
                toServer.write(len);
                toServer.write(nomefile_utf8);

                // effettuo encoding di stringa in UTF-8 e calcolo la
                // dimensione in byte del buffer risultante
                var stringa_utf8 = stringa.getBytes("UTF-8");
                var stringa_len = stringa_utf8.length;
                
                // effettuo encoding della dimensione del buffer stringa_utf8
                // in un unsigned a 16 bit con ordine big endian
                len[0] = (byte)((stringa_len & 0xFF00) >> 8);
                len[1] = (byte)(stringa_len & 0xFF);

                // trasmetto stringa
                toServer.write(len);
                toServer.write(stringa_utf8);

                /* Leggo dimensione messaggio di risposta */
                fromServer.read(len);
                var to_read = ((int)(len[0] & 0xFF) << 8) | (int)(len[1] & 0xFF);

                /* Stampo contenuto risposta a video */
                while (to_read > 0) {
                    var buffer = new byte[4096];

                    var bufsize = buffer.length;
                    var sz = (to_read < bufsize) ? to_read : bufsize;

                    var nread = fromServer.read(buffer, 0, sz);
                    System.out.write(buffer, 0, nread);

                    to_read -= nread;
                }

            } while(true);

            s.close();
        }
        catch(IOException e){
            System.err.println(e.getMessage());
            e.printStackTrace();
            System.exit(100);
        }
    }
}

